### PR TITLE
[ALLUXIO-2514]Fix bug of set ttl to directory 

### PR DESCRIPTION
**Pull request by gjhkael**
https://alluxio.atlassian.net/browse/ALLUXIO-2514

### CODE REVIEW

**Comment by gjhkael** [comment]
@apc999  jsimsa Another problem, when user create directory with CreateDirectoryOptions' ttl >0, actually, it does not set ttl to the directory. So,  It is necessary to adjust method of createDirectory in FileSystemMaster to support set ttl.

File: core/server/src/main/java/alluxio/master/file/meta/Inode.java [hunk 1]
```diff
@@ -471,7 +471,7 @@ public boolean equals(Object o) {
   protected Objects.ToStringHelper toStringHelper() {
     return Objects.toStringHelper(this).add("id", mId).add("name", mName).add("parentId", mParentId)
         .add("creationTimeMs", mCreationTimeMs).add("pinned", mPinned).add("deleted", mDeleted)
-        .add("ttl", mTtl).add("mTtlAction", mTtlAction)
+        .add("ttl", mTtl).add("TtlAction", mTtlAction)
```
Comment (apc999) [inline core/server/src/main/java/alluxio/master/file/meta/Inode.java]: Camel case please `ttlAction`

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 2]
```diff
@@ -786,6 +813,30 @@ public void setTtlForFileWithFreeOperationTest() throws Exception {
...
+            CreateDirectoryOptions.defaults().setRecursive(true);
+    mFileSystemMaster.createDirectory(NESTED_URI, createDirectoryOptions);
+    long blockId = createFileWithSingleBlock(NESTED_FILE_URI);
+    Assert.assertEquals(1, mBlockMaster.getBlockInfo(blockId).getLocations().size());
+    // Set ttl & operation
+    SetAttributeOptions options = SetAttributeOptions.defaults();
+    options.setTtl(0);
+    options.setTtlAction(TtlAction.FREE);
+    mFileSystemMaster.setAttribute(NESTED_URI, options);
+    Command heartBeat =
```
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: `heartbeat`

Comment (gjhkael) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: you mean the variate heartBeat rename to heartbeat?

Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: yup, for consistency with the rest of the code base (e.g. `workerHeartbeat`)

File: core/server/src/main/java/alluxio/master/file/meta/Inode.java (hunk 1 above)
Comment (gjhkael) [inline core/server/src/main/java/alluxio/master/file/meta/Inode.java]: done

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java (hunk 2 above)
Comment (gjhkael) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: done

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 3]
```diff
@@ -1702,12 +1709,12 @@ private long journalCreatePathResult(InodeTree.CreatePathResult createResult) {
     long counter = AsyncJournalWriter.INVALID_FLUSH_COUNTER;
     for (Inode<?> inode : createResult.getModified()) {
       InodeLastModificationTimeEntry inodeLastModificationTime =
-          InodeLastModificationTimeEntry.newBuilder()
-          .setId(inode.getId())
-          .setLastModificationTimeMs(inode.getLastModificationTimeMs())
-          .build();
+              InodeLastModificationTimeEntry.newBuilder()
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: undo this formatting change

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 4]
```diff
@@ -1722,10 +1729,10 @@ private long journalCreatePathResult(InodeTree.CreatePathResult createResult) {
     }
     for (Inode<?> inode : createResult.getPersisted()) {
       PersistDirectoryEntry persistDirectory = PersistDirectoryEntry.newBuilder()
-          .setId(inode.getId())
-          .build();
+              .setId(inode.getId())
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: ditto

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 5]
```diff
@@ -1682,7 +1682,14 @@ private long createDirectoryAndJournal(LockedInodePath inodePath, CreateDirector
       CreateDirectoryOptions options) throws InvalidPathException, FileAlreadyExistsException,
       IOException, AccessControlException, FileDoesNotExistException {
     try {
-      return mInodeTree.createPath(inodePath, options);
+      InodeTree.CreatePathResult createResult = mInodeTree.createPath(inodePath, options);
+      InodeDirectory inodeDirectory = (InodeDirectory) inodePath.getInode();
+      // If inodeDirectory's ttl not equals Constants.NO_TTL, it should insert into mTtlBuckets
+      if (createResult.getCreated().size() > 0) {
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: I think we also need to add this logic to journal replay (see the InodeFileEntry if branch in the processJournalEntry) ... there should be a unit test for this as well

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 3 above)
Comment (gjhkael) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: done

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 4 above)
Comment (gjhkael) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: done

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 5 above)
Comment (gjhkael) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: Add journal replay, but how to design the unit test?

**Comment by gjhkael** [comment]
Sorry, what PTAL means?

**Comment by apc999** [comment]
@gjhkael  could you resolve the conflict?

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 5 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: the unit test would do the following:

1) use manual scheduling for TTL
2) request directory creation with TTL set
3) stop block master and file system master
4) start block master and file system master again with the same journal
5) manually schedule TTL and check the directory has been freed / deleted

I am happy to add this test in a followup PR myself as the setup is non-trivial.

Comment (gjhkael) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: Ok, thanks for your helps. Could link your PR in this pull request, I can learn about it.

**Comment by yupeng9** [comment]
@jsimsa Has your comment been addressed by the latest commit?

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 6]
```diff
@@ -725,6 +726,21 @@ public void createFileWithTtl() throws Exception {
...
 
   /**
+   * Tests that an exception is in the
+   * {@link FileSystemMaster#createDirectory(AlluxioURI, CreateDirectoryOptions)} with a TTL
+   * set in the {@link CreateDirectoryOptions} after the TTL check was done once.
+   */
+  @Test
+  public void createDirectoryWithTtl() throws Exception {
+    CreateDirectoryOptions directoryOptions =
+            CreateDirectoryOptions.defaults().setRecursive(true).setTtl(0);
```
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: this is not the correct indentation (Alluxio Java style guideline is to use 4 spaces), please fix this through all the newly contributed code

**Comment by jsimsa** [comment]
@yupeng9 yes, but I noticed that the newly contributed code is not correctly indented

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 7]
```diff
@@ -66,6 +66,7 @@
 import alluxio.master.journal.JournalFactory;
 import alluxio.master.journal.JournalOutputStream;
 import alluxio.metrics.MetricsSystem;
+import alluxio.proto.journal.File;
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: replace this with `alluxio.proto.journal.File.InodeDirectoryEntry`

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 8]
```diff
@@ -313,6 +314,12 @@ public void processJournalEntry(JournalEntry entry) throws IOException {
       }
     } else if (entry.hasInodeDirectory()) {
       try {
+        // Add the directory to TTL buckets, the insert automatically rejects directory
+        // w/ Constants.NO_TTL
+        File.InodeDirectoryEntry inodeDirectoryEntry = entry.getInodeDirectory();
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: replace `File.InodeDirectoryEntry` with just `InodeDirectory` ... this is for consistency with how other journal entry imports

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 9]
```diff
@@ -745,6 +761,32 @@ public void setTtlForFileWithNoTtl() throws Exception {
   }
 
   /**
+   * Tests that an exception is thrown when trying to get information about a Directory after
+   * it has been deleted because of a TTL of 0.
+   */
+  @Test
+  public void setTtlForDirectoryWithNoTtl() throws Exception {
+    CreateDirectoryOptions createDirectoryOptions =
+            CreateDirectoryOptions.defaults().setRecursive(true);
```
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: indent

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 10]
```diff
@@ -745,6 +761,32 @@ public void setTtlForFileWithNoTtl() throws Exception {
...
+    mFileSystemMaster.createDirectory(NESTED_URI, createDirectoryOptions);
+    mFileSystemMaster.createDirectory(NESTED_DIR_URI, createDirectoryOptions);
+    CreateFileOptions createFileOptions =
+            CreateFileOptions.defaults().setBlockSizeBytes(Constants.KB).setRecursive(true);
```
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: fix indent

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 11]
```diff
@@ -765,6 +807,24 @@ public void setSmallerTtlForFileWithTtl() throws Exception {
   }
 
   /**
+   * Tests that an exception is thrown when trying to get information about a Directory after
+   * it has been deleted after the TTL has been set to 0.
+   */
+  @Test
+  public void setSmallerTtlForDirectoryWithTtl() throws Exception {
+    CreateDirectoryOptions createDirectoryOptions =
+            CreateDirectoryOptions.defaults().setRecursive(true).setTtl(Constants.HOUR_MS);
```
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: fix indent

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 12]
```diff
@@ -777,11 +837,35 @@ public void setTtlForFileWithFreeOperationTest() throws Exception {
...
+  }
+
+  /**
+   * Tests that file information is still present after it has been freed after the parent
+   * directory's TTL has been set to 0.
+   */
+  @Test
+  public void setTtlForDirectoryWithFreeOperationTest() throws Exception {
+    CreateDirectoryOptions createDirectoryOptions =
+            CreateDirectoryOptions.defaults().setRecursive(true);
```
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: fix indent

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 13]
```diff
@@ -777,11 +837,35 @@ public void setTtlForFileWithFreeOperationTest() throws Exception {
...
+    mFileSystemMaster.createDirectory(NESTED_URI, createDirectoryOptions);
+    long blockId = createFileWithSingleBlock(NESTED_FILE_URI);
+    Assert.assertEquals(1, mBlockMaster.getBlockInfo(blockId).getLocations().size());
+    // Set ttl & operation
+    SetAttributeOptions options = SetAttributeOptions.defaults();
+    options.setTtl(0);
+    options.setTtlAction(TtlAction.FREE);
+    mFileSystemMaster.setAttribute(NESTED_URI, options);
+    Command heartbeat =
+            mBlockMaster.workerHeartbeat(mWorkerId1, ImmutableMap.of("MEM", (long) Constants.KB),
```
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: fix indent

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 14]
```diff
@@ -777,11 +837,35 @@ public void setTtlForFileWithFreeOperationTest() throws Exception {
...
+                    ImmutableList.of(blockId), ImmutableMap.<String, List<Long>>of());
```
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: fix indent

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 15]
```diff
@@ -803,6 +887,21 @@ public void setLargerTtlForFileWithTtl() throws Exception {
   }
 
   /**
+   * Tests that a directory has not been deleted after the TTL has been reset to a valid value.
+   */
+  @Test
+  public void setLargerTtlForDirectoryWithTtl() throws Exception {
+    CreateDirectoryOptions createDirectoryOptions =
+            CreateDirectoryOptions.defaults().setRecursive(true).setTtl(0);
```
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: fix indent

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 16]
```diff
@@ -803,6 +887,21 @@ public void setLargerTtlForFileWithTtl() throws Exception {
...
+    mFileSystemMaster.createDirectory(NESTED_URI, createDirectoryOptions);
+    mFileSystemMaster.setAttribute(NESTED_URI,
+            SetAttributeOptions.defaults().setTtl(Constants.HOUR_MS));
```
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: fix indent

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 17]
```diff
@@ -819,6 +918,23 @@ public void setNoTtlForFileWithTtl() throws Exception {
   }
 
   /**
+   * Tests that the original TTL is removed after setting it to {@link Constants#NO_TTL} for
+   * a directory.
+   */
+  @Test
+  public void setNoTtlForDirectoryWithTtl() throws Exception {
+    CreateDirectoryOptions createDirectoryOptions =
+            CreateDirectoryOptions.defaults().setRecursive(true).setTtl(0);
```
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: fix indent

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 18]
```diff
@@ -819,6 +918,23 @@ public void setNoTtlForFileWithTtl() throws Exception {
...
+    mFileSystemMaster.createDirectory(NESTED_URI, createDirectoryOptions);
+    // After setting TTL to NO_TTL, the original TTL will be removed, and the file will not be
+    // deleted during next TTL check.
+    mFileSystemMaster.setAttribute(NESTED_URI,
+            SetAttributeOptions.defaults().setTtl(Constants.NO_TTL));
```
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: fix indent

**Comment by apc999** [comment]
LGTM

**Comment by jsimsa** [comment]
@gjhkael FYI here is the PR for adding unit tests for journal replay (https://github.com/Alluxio/alluxio/pull/4624)

**Pull request merged**